com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java

```diff
@@ -1425,6 +1432,89 @@ private void configurePreBuildProjectOptions() throws IOException, CompileExcept
         }
     }
 
+    private void transpileLua(IProgress monitor) throws CompileExceptionError, IOException {
+        List<ILuaTranspiler> transpilers = PluginScanner.getOrCreatePlugins("com.defold.extension.pipeline", ILuaTranspiler.class);
+        if (transpilers != null) {
+            IProgress transpilerProgress = monitor.subProgress(1);
+            transpilerProgress.beginTask("Transpiling to Lua", 1);
+            for (ILuaTranspiler transpiler : transpilers) {
+                IResource buildFileResource = getResource(transpiler.getBuildFileResourcePath());
+                if (buildFileResource.exists()) {
+                    String ext = "." + transpiler.getSourceExt();
+                    List<IResource> sources = inputs.stream()
+                            .filter(s -> s.endsWith(ext))
+                            .map(this::getResource)
+                            .collect(Collectors.toList());
+                    if (!sources.isEmpty()) {
+                        boolean useProjectDir = buildFileResource instanceof DefaultResource && sources.stream().allMatch(s -> s instanceof DefaultResource);
+                        File sourceDir;
+                        if (useProjectDir) {
+                            sourceDir = new File(rootDirectory);
+                        } else {
+                            sourceDir = Files.createTempDirectory("tr-" + transpiler.getClass().getSimpleName()).toFile();
+                            buildFileResource.getContent();
+                            File buildFile = new File(sourceDir, buildFileResource.getPath());
+                            Path buildFilePath = buildFile.toPath();
+                            Files.write(buildFilePath, buildFileResource.getContent());
+                            Files.setLastModifiedTime(buildFilePath, FileTime.fromMillis(buildFileResource.getLastModified()));
+                            for (IResource source : sources) {
+                                Path sourcePath = new File(sourceDir, source.getPath()).toPath();
+                                Files.write(sourcePath, source.getContent());
+                                Files.setLastModifiedTime(sourcePath, FileTime.fromMillis(source.getLastModified()));
+                            }
+                        }
+                        File outputDir = new File(rootDirectory, "build/tr/" + transpiler.getClass().getSimpleName());
```

Comment:
The classloader will load only one class of the same name, and which one would it be is undefined. It's not allowed to use the same class name for multiple transpiler extensions, so there is no issue here.

Reply:
Just because the behavior in the JVM is undefined, it doesn't mean there is no issue here.

I realize we want to be able to reload these classes in the editor, but perhaps we could verify that the class we're loading corresponds to the same location on disk as the one we're replacing?

It doesn't have to handle the situation gracefully, but throwing an exception is better than just silently producing unexpected build output. This could be really tricky for the users to track down if it happens.

Reply:
> I realize we want to be able to reload these classes in the editor

Not really, updating will require editor restart, same as with other extensions like lua preprocessors and spine.